src/dbnode/storage/bootstrap/bootstrapper/commitlog/source.go

```diff
@@ -701,6 +739,106 @@ func (s *commitLogSource) mostRecentCompleteSnapshotByBlockShard(
 	return mostRecentSnapshotsByBlockShard
 }
 
+type indexSnapshot struct {
+	fileSet   fs.FileSetFile
+	info      index.IndexVolumeInfo
+	fulfilled result.ShardTimeRanges
+}
+
+// mostRecentCompleteSnapshotByBlock returns a
+// map[xtime.UnixNano]fs.FileSetFile with the contract that
+// for each block in shardsTimeRanges, an entry will
+// exist in the map such that FileSetFile.CachedSnapshotTime is the
+// actual cached snapshot time, or the blockStart.
+func (s *commitLogSource) mostRecentCompleteIndexSnapshotByBlock(
+	shardsTimeRanges result.ShardTimeRanges,
+	blockSize time.Duration,
+	indexSnapshotFiles fs.FileSetFilesSlice,
+	indexInfoFiles []fs.ReadIndexInfoFileResult,
+	fsOpts fs.Options,
+) map[xtime.UnixNano]indexSnapshot {
+	var (
+		minBlock, maxBlock                = shardsTimeRanges.MinMax()
+		mostRecentIndexSnapshotsByBlock   = map[xtime.UnixNano]indexSnapshot{}
+		latestIndexVolumeInfoByBlockStart = map[xtime.UnixNano]index.IndexVolumeInfo{}
+	)
+
+	for _, info := range indexInfoFiles {
+		latestIndexVolumeInfoByBlockStart[xtime.ToUnixNano(info.ID.BlockStart)] = info.Info
+	}
+
+	for currBlockStart := minBlock.Truncate(blockSize); currBlockStart.Before(maxBlock); currBlockStart = currBlockStart.Add(blockSize) {
+		// Anonymous func for easier clean up using defer.
+		func() {
+			var (
+				currBlockUnixNanos = xtime.ToUnixNano(currBlockStart)
+				mostRecentSnapshot fs.FileSetFile
+				indexVolumeInfo    index.IndexVolumeInfo
+				ok                 bool
+			)
+
+			defer func() {
+				if mostRecentSnapshot.IsZero() {
+					// If we were unable to determine the most recent snapshot time for a given
+					// blockStart, then just fall back to using the blockStart time as that will
+					// force us to read the entire commit log for that duration.
+					mostRecentSnapshot.CachedSnapshotTime = currBlockStart
```

Comment:
As discussed, let's remove cached snapshot time and just rely on whether present in map or not?